backend/x11/backend.c

```diff
@@ -611,43 +619,30 @@ struct wlr_backend *wlr_x11_backend_create(struct wl_display *display,
 			wlr_log(WLR_ERROR, "Failed to query DRI3 DRM FD");
 			goto error_event;
 		}
+	}
 
-		char *drm_name = drmGetDeviceNameFromFd2(x11->drm_fd);
-		wlr_log(WLR_DEBUG, "Using DRM node %s", drm_name);
-		free(drm_name);
+	x11->renderer = wlr_renderer_autocreate(&x11->backend);
+	if (x11->renderer == NULL) {
+		wlr_log(WLR_ERROR, "Failed to create renderer");
+		goto error_event;
+	}
 
-		int drm_fd = fcntl(x11->drm_fd, F_DUPFD_CLOEXEC, 0);
-		if (drm_fd < 0) {
-			wlr_log(WLR_ERROR, "fcntl(F_DUPFD_CLOEXEC) failed");
-			goto error_event;
-		}
+	uint32_t caps = renderer_get_render_buffer_caps(x11->renderer);
 
-		struct wlr_gbm_allocator *gbm_alloc = wlr_gbm_allocator_create(drm_fd);
-		if (gbm_alloc == NULL) {
-			wlr_log(WLR_ERROR, "Failed to create GBM allocator");
-			close(drm_fd);
-			goto error_event;
-		}
-		x11->allocator = &gbm_alloc->base;
+	const struct wlr_drm_format_set *pixmap_formats;
+	if (x11->have_dri3 && (caps & WLR_BUFFER_CAP_DMABUF)) {
 		pixmap_formats = &x11->dri3_formats;
-	} else if (x11->have_shm) {
-		x11->drm_fd = -1;
-		struct wlr_shm_allocator *shm_alloc = wlr_shm_allocator_create();
-		if (shm_alloc == NULL) {
-			wlr_log(WLR_ERROR, "Failed to create shared memory allocator");
-			goto error_event;
-		}
-		x11->allocator = &shm_alloc->base;
+	} else if (x11->have_shm && (caps & WLR_BUFFER_CAP_DATA_PTR)) {
```

Comment:
The X11 backend needs `WLR_BUFFER_CAP_SHM`, it doesn't use/need/care about `CAP_DATA_PTR`

Reply:
Those are renderer caps, pixman gives `CAP_DATA_PTR`.

Reply:
Oh right. Hm, I guess that doesn't map that well, we could have buffers with `get_data_ptr` but not `get_shm`, like `wlr_wl_shm_buffer` in https://github.com/swaywm/wlroots/pull/2892. Not a big deal for now since allocators are hard-coded and no allocator will return such a buffer.